File: CS-IOS/Classes/Core/Classes/Network/CSMultiResponse.m
//
//  Created by Rene Dohan on 12/29/12.
//


#import "CSMultiResponse.h"
#import "NSObject+CSExtension.h"


@implementation CSMultiResponse {
    CSResponse *_request;
}

- (instancetype)construct:(CSResponse *)request {
    super.construct;
    [self add:request];
    return self;
}

- (CSResponse *)add:(CSResponse *)request {
    _request = request;
    [self failIfFail:request];
    return request;
}

- (void)cancel {
    [super cancel];
    [_request cancel];
}

- (CSResponse *)addLast:(CSResponse *)request {
    return [[self add:request] onSuccess:^(id id) {
        [self success:id];
    }];
}

@end

File: Renetik/Renetik/Classes/Core/Classes/Response/CSMultiResponse.swift
//
// Created by Rene Dohan on 3/9/20.
//

import RenetikObjc

public class CSMultiResponse: CSResponse<NSMutableArray> {
    var response: CSResponseProtocol!

    override public init() {
        super.init(); data = NSMutableArray()
    }

    @discardableResult
    public func add<Data: AnyObject, Response: CSResponse<Data>>(_ response: Response) -> Response {
        data.add(response.data)
        self.response = failIfFail(response)
        return response
    }

    @discardableResult
    public func add<Data: AnyObject, Response: CSResponse<Data>>(last request: Response) -> Response {
        add(request).onSuccess { _ in self.success() }
    }

    override public func cancel() {
        super.cancel()
        response.cancel()
    }
}
